sdks/python/apache_beam/examples/complete/game/game_stats.py

```diff
@@ -0,0 +1,285 @@
+publishing.
+
+To run the Java injector:
+<beam_root>/examples/java8$ mvn compile exec:java \
+    -Dexec.mainClass=org.apache.beam.examples.complete.game.injector.Injector \
+    -Dexec.args="$PROJECT_ID $PUBSUB_TOPIC none"
+
+For a description of the usage and options, use -h or --help.
+
+To specify a different runner:
+  --runner YOUR_RUNNER
+
+NOTE: When specifying a different runner, additional runner-specific options
+      may have to be passed in as well
+NOTE: With DataflowRunner, the --setup_file flag must be specified to handle the
+      'util' module
+
+EXAMPLES
+--------
+
+# DirectRunner
+python game_stats.py \
+    --project $PROJECT_ID \
+    --topic projects/$PROJECT_ID/topics/$PUBSUB_TOPIC \
+    --dataset $BIGQUERY_DATASET
+
+# DataflowRunner
+python game_stats.py \
+    --project $PROJECT_ID \
+    --topic projects/$PROJECT_ID/topics/$PUBSUB_TOPIC \
+    --dataset $BIGQUERY_DATASET \
+    --runner DataflowRunner \
+    --setup_file ./setup.py \
+    --staging_location gs://$BUCKET/user_score/staging \
+    --temp_location gs://$BUCKET/user_score/temp
+"""
+
+from __future__ import absolute_import
+
+import argparse
+import logging
+
+import apache_beam as beam
+from apache_beam.transforms import combiners
+
+from apache_beam.examples.complete.game.util import util
+
+
+class CalculateSpammyUsers(beam.PTransform):
+  """Filter out all but those users with a high clickrate, which we will
+  consider as 'spammy' uesrs.
+
+  We do this by finding the mean total score per user, then using that
+  information as a side input to filter out all but those user scores that are
+  larger than (mean * SCORE_WEIGHT).
+  """
+  SCORE_WEIGHT = 2.5
+
+  def expand(self, user_scores):
+    # Get the sum of scores for each user.
+    sum_scores = (
+        user_scores
+        | 'SumUsersScores' >> beam.CombinePerKey(sum)
+    )
+
+    # Extract the score from each element, and use it to find the global mean.
+    global_mean_score = (
+        sum_scores
+        | beam.Values()
+        # NOTE: DataflowRunner currently does not support combiners
+        | beam.CombineGlobally(combiners.MeanCombineFn()).as_singleton_view()
+    )
+
+    # Filter the user sums using the global mean.
+    filtered = (
+        sum_scores
+        # Use the derived mean total score (global_mean_score) as a side input.
+        | 'ProcessAndFilter' >> beam.Filter(
+            lambda (_, score), global_mean:\
+                score > global_mean * self.SCORE_WEIGHT,
+            global_mean_score)
+    )
+    return filtered
+
+
+class UserSessionActivity(beam.DoFn):
+  """Calculate and output an element's session duration, in seconds."""
+  def process(self, elem, window=beam.DoFn.WindowParam):
+    yield (window.end.micros - window.start.micros) / 1000000
+
+
+def run(argv=None):
+  """Main entry point; defines and runs the hourly_team_score pipeline."""
+  parser = argparse.ArgumentParser()
+
+  parser.add_argument('--project',
```

Comment:
<!--thread_id:cc_128809164_t; commit:95d457e2ad40bb613cb8abda68903ad026f0cccd; resolved:0-->
<!--section:context-quote-->
> **davidcavazos** wrote:
> You still need ``--project`` even if you are running on the DirectRunner for the BigQuery dataset output. The user could use only ``--topic`` and ``--dataset``, this would validate correctly but crash with an exception.

<!--section:body-->
Defining the same flag twice would have unexpected results, because both flags will be added to argparse. Could you use the existing flag and validate that (assert) that it is defined before running?